raft/src/main/java/org/apache/kafka/raft/RaftClient.java

```diff
@@ -24,14 +24,37 @@
 public interface RaftClient {
 
     /**
-     * Initialize the state machine that will be used by this client. This should
-     * only be called once after creation and calls to {@link RecordAppender#append(Records)} should
-     * be made until this method returns.
-     *
-     * @param stateMachine The state machine implementation
+     * Initialize the client and provide a listener to be notified of election events.
+     * @param listener
      * @throws IOException For any IO errors during initialization
      */
-    void initialize(ReplicatedStateMachine stateMachine) throws IOException;
+    void initialize(ElectionListener listener) throws IOException;
```

Comment:
I remember that we had a discussion about callbacks APIs `ElectionListener` vs `CompletableFuture` APIs. I don't have a strong opinion of one over the other but it would be nice if the API used one.

I understand that the issue is that `ElectionListener` is more like an async stream since it can fire multiple times with different values. While `append` and `read` can only return one value.

I can think of two solutions:
1. CompletableFuture<LeaderState> leaderState(LeaderState knownState) - the returned future completes if/when the leader state is different from the known state. The known state can something like `Optional<Int>` for optional epoch. The first time it is called the state machine can assume it is a follower by sending `Optional.empty()`.
2. Use Java's Flow: https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/Flow.html